### DISCUSSION TITLE
How can I use these, getOrCreateDocument and initializeWorkspace in CLI?

### DISCUSSION THREAD

**Discussion by negahama** [eclipse-langium/langium#1700] [Q&A] [answered]
Hello everyone
I made my CLI, It works well, but It has a big serious problem. It's too slow.
Most of the time is spent on `langiumDocuments.getOrCreateDocument()`. 99% of the total working time is spent here really.

And I made My CLI base on #861, so every time my CLI transpile the files, It calls `workspace.WorkspaceManager.initializeWorkspace()`

I'm wondering why `getOrCreateDocument()` is taking so much time, but I'm more curious about how to solve it than why.
How could I not call `initializeWorkspace()` every time for performance?
Thanks.

**Comment by msujew** [comment]
Hey @negahama,

Can you share some of your code with us? Without more info, it's very hard to figure out how to fix your performance issue. Have you tried profiling your CLI yet? There are often times performance bottlenecks im the scoping and validation implementations.

**Reply by negahama** [reply]
Hi @msujew 

My project is [Here](https://github.com/negahama/NegahamaOrg.ScalaScript). You can try to it
```
npm install
cd scala-script-cli
npm run all
```
then my cli program will generate files to `scala-script-cli/samples/generated/` from the sample files in `scala-script-cli/samples/`
You could see the result message in terminal in a minute
As I told you, the most of time you wait is spent `getOrCreateDocument()`
Thank you for help

**Reply by msujew** [reply, marked as answer]
Looking at this through a profiler, it seems like the parser is taking a very long time on your files - mainly with attempting to perform the lookahead. This indicates very long common prefixes in the grammar:

![grafik](https://github.com/user-attachments/assets/5bba8582-2c1d-4e11-a969-cc966010ed86)

Looking at your grammar, it seems like your language features a lot of ambiguity in the first few tokens. For example `ForOf` and `ForTo` can be joined into a single rule, and everything that starts with `(Annotation)? (export?='export')? (private?='private')? (static?='static')?` can be moved into a single prefixed rule that then fans out into different branches. Also, the lookahead creates a warning in your code saying that `Block` is completely ambiguous:

```
Ambiguous Alternatives Detected: <0, 1> in <OR1> inside <Block​> Rule,
<{, }, }> may appears as a prefix path in all these alternatives.
See: https://chevrotain.io/docs/guide/resolving_grammar_errors.html#AMBIGUOUS_ALTERNATIVES
For Further details.
```

**Comment by negahama** [comment]
@msujew Thank you so much...
I have something I want to make clear.
I thought `getOrCreateDocument()` have nothing to do with parsing... Is it right?
Are Lexing and parsing processed in `DocumentFactory.build`?

**Reply by msujew** [reply]
Creating a document automatically parses it's content. There is no document without content.

**Reply by negahama** [reply]
I see... that is why `getOrCreateDocument` takes so long time... thank you!!

And one more...
Whenever I want to build, Do I have to call `initializeWorkspace` in CLI?
Is there better way to do that?

**Reply by msujew** [reply]
I would recommend to use the `DocumentBuilder.build` API to build documents. The WorkspaceManager is using that API internally as well.
